### DISCUSSION TITLE
Can we have useCreateTestCtx?

### DISCUSSION THREAD

**Discussion by MOZGIII** [reatom/reatom#994] [Ideas]
I've been doing some development with reatom, and figured we need a way to create a test ctx on a per-page basis.
We have some dev-routes with pages that need mocking - sort of like storybooks but at the same app code as the rest of the app.

```ts
import { Fn } from "@reatom/framework";
import { createTestCtx, TestCtx } from "@reatom/testing";
import { useRef } from "react";

export const useCreateTestCtx = (extension?: Fn<[TestCtx]>) => {
  const ctxRef = useRef<null | TestCtx>(null);
  if (!ctxRef.current) {
    ctxRef.current = createTestCtx();
    extension?.(ctxRef.current);
  }
  return ctxRef.current;
};
```

It would also be fine to have a just more generic way to create and destroy the contexts, so that `useCreateCtx` doesn't invoke the the `createCtx` itself but rather allows the user to do so.

```ts
import { Ctx, Fn } from "@reatom/framework";
import { useRef } from "react";

export const useCreateCustomCtx = <T extends Ctx>(create: Fn<[], T>) => {
  const ctxRef = useRef<null | T>(null);
  if (!ctxRef.current) {
    ctxRef.current = create();
  }
  return ctxRef.current;
};
```

The second variant should work better as it does not have direct dependency on both `react` and a `@reatom/testing` - which would make it quite puzzling where to place this code.

**Comment by artalar** [comment]
Hi! Nice idea, created #1012
